### PR TITLE
Allow Models-config to change (configReload) while deployments are in progress

### PR DESCRIPTION
**Pull request by maheshkelkar**
- Added ModelConfig model, DB table, etc to keep track of old configs.

- Added checksum calculations to keep track of each configReload
  iteration. This calculation reads all the directories and files
  in a deterministic order and computes checksum, which should remain
  same for given config on any platform.

- The checksum is stored in the DeployDB app object as well Flow object
  in order to associate them with the given iteration of config.

- On the subsequent Flow/Deployment transitions, if checksum from the
  Flow matches with that in the App object, then in-memory (i.e. registry)
  Model objects are used. Otherwise Model objects are read from the
  ModelConfig objects.

- In loadConfigModels, moved code repetitions such as loading from
  string, inserting into registry, checksum calculations, etc.
  to common function

- Added cucumber and spock tests to validate the code.

References #126

### CODE REVIEW

**Comment by maheshkelkar** [comment]
@GiriDandu Please review.

File: features/api/config.feature [hunk 1]
```diff
@@ -16,14 +16,143 @@ Feature: DeployDB config APIs
...
     Then the response should be 200
     And the response body should be:
     """
-    Failed: Configuration reload is not allowed while deployments are in progress
+    Done!
+    """
+
+
+  @freezetime
+  Scenario: If deployment is CREATED state and configuration is reloaded, then
```
Comment (rtyler) [inline features/api/config.feature]: This scenario name is too long. Can this be summarized into less than 80 chars :p 

Test reporters will all truncate this after the first line which will make the reports not very useful

File: features/api/config.feature [hunk 2]
```diff
@@ -16,14 +16,143 @@ Feature: DeployDB config APIs
...
+          "sourceUrl" : "http://example.com/maven/com.example.cucumber/cucumber-artifact/1.0.1/cucumber-artifact-1.0.1.jar",
+          "createdAt" : "{{created_timestamp}}"
+        },
+        "status" : "STARTED",
+        "service" : "faas",
+        "environment" : "pre-prod",
+        "createdAt" : "{{created_timestamp}}"
+      }
+    """
+    And the webhook 2 should be invoked with the JSON:
```
Comment (rtyler) [inline features/api/config.feature]: I don't think there's any need to test the second webhook here right? We only need to verify that tihngs work the same for one of the webhooks that was configured right?

File: features/api/config.feature [hunk 3]
```diff
@@ -16,14 +16,143 @@ Feature: DeployDB config APIs
...
+        "status" : "STARTED",
+        "service" : "faas",
+        "environment" : "pre-prod",
+        "createdAt" : "{{created_timestamp}}"
+      }
+    """
+
+
+  @freezetime
+  Scenario: If deployment is STARTED state and configuration is reloaded, then
```
Comment (rtyler) [inline features/api/config.feature]: ETOOLONG

File: src/cucumber/groovy/deploydb/cucumber/StubAppRunner.groovy [hunk 4]
```diff
@@ -90,12 +82,14 @@ public class StubAppRunner<C extends Configuration> {
...
+                            workFlow = application.workFlow
+
+                            /**
+                             * Setup config checksum for tests
                              */
-                            serviceRegistry = application.workFlow.serviceRegistry
-                            environmentRegistry = application.workFlow.environmentRegistry
-                            promotionRegistry = application.workFlow.promotionRegistry
-                            pipelineRegistry = application.workFlow.pipelineRegistry
+                            application.configChecksum = "0xdead"
```
Comment (GiriDandu) [inline src/cucumber/groovy/deploydb/cucumber/StubAppRunner.groovy]: Not sure the reason for assignment.  Don't we calculate this and store it?

File: features/api/config.feature (hunk 1 above)
Comment (maheshkelkar) [inline features/api/config.feature]: ok. will do.

File: src/cucumber/groovy/step_definitions/DeploymentSteps.groovy [hunk 5]
```diff
@@ -36,7 +36,7 @@ Given(~/^there is a deployment$/) { ->
         d1.addPromotionResult(p1)
 
         /* Create a flow */
-        Flow f = new Flow(a1, "faas")
+        Flow f = new Flow(a1, "faas", "0xdead")
```
Comment (GiriDandu) [inline src/cucumber/groovy/step_definitions/DeploymentSteps.groovy]: Is this because we don't have config in the DeploymentSteps?

File: features/api/config.feature (hunk 2 above)
Comment (maheshkelkar) [inline features/api/config.feature]: Yes. We don't. After adding one, I added another just to be sure. I can take that out.

File: features/api/config.feature (hunk 3 above)
Comment (maheshkelkar) [inline features/api/config.feature]: fixing

File: src/cucumber/groovy/deploydb/cucumber/StubAppRunner.groovy (hunk 4 above)
Comment (maheshkelkar) [inline src/cucumber/groovy/deploydb/cucumber/StubAppRunner.groovy]: For most of the cucumber tests, we do not load config, because it affects our test scenarios. So hardcoding it here, so that other checks performed on it don't fail

File: src/cucumber/groovy/step_definitions/WebhookSteps.groovy [hunk 6]
```diff
@@ -5,28 +5,48 @@ import com.fasterxml.jackson.core.JsonToken
...
+                workFlow.deployDBApp.configChecksum, configBody,
+                workFlow.defaultIdent, ModelType.WEBHOOK)
+
+        withSession {
+            ModelConfigDAO modelConfigDAO = new ModelConfigDAO(sessionFactory)
+            modelConfigDAO.persist(modelConfig)
+        }
+    }
+
+    withTestWebhookServer { RequestWebhookObject requestWebhookObject ->
```
Comment (GiriDandu) [inline src/cucumber/groovy/step_definitions/WebhookSteps.groovy]: Not too sure why this change is done.  Java docs to explain it?

File: src/cucumber/groovy/step_definitions/WebhookSteps.groovy [hunk 7]
```diff
@@ -34,18 +54,11 @@ Given(~/^a (.*?) webhook "(.*?)" configuration:$/) { String webhookType,
...
-        ModelLoader<Webhook> webhookLoader = new ModelLoader<>(Webhook.class)
-        webhookManager.webhook = webhookLoader.loadFromString(configBody)
-
-        /*
          * Set the content type from the webhook and the event type. The content type will be
          * checked when deploydb invokes webhooks
          */
         requestWebhookObject.contentTypeParam = "application/vnd.deploydb."+
-                                                 webhookType+eventType+".v1+json"
+                webhookType+eventType+".v1+json"
```
Comment (GiriDandu) [inline src/cucumber/groovy/step_definitions/WebhookSteps.groovy]: Not a big fan of this indenting. Also can you stick the code changes that you state in PR ;-) Open a new PR for code cleanup or say that some code cleanup is also in this PR.

File: src/cucumber/groovy/step_definitions/WebhookSteps.groovy [hunk 8]
```diff
@@ -5,28 +5,48 @@ import com.fasterxml.jackson.core.JsonToken
...
 import webhookTestServer.models.RequestWebhookObject
 
+
 Given(~/^a (.*?) webhook "(.*?)" configuration:$/) { String webhookType,
                                                      String eventType, String configBody ->
 
     List<String> paths = getUrlPathFromWebhookConfigBody(configBody, eventType)
+    withWorkFlow { WorkFlow workFlow ->
+        /*
+         * Load the webhook configuration in webhookManager
```
Comment (GiriDandu) [inline src/cucumber/groovy/step_definitions/WebhookSteps.groovy]: Fix the comment because you don't have webhookManager anymore

File: src/main/groovy/db/migration/V8__create_configs_table.groovy [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        /* Sql commands */
+        List<String> commands = []
+
+        /*
+         * Add modelConfigs table
+         */
+        commands += """
+            CREATE TABLE modelConfigs (
+                id BIGINT AUTO_INCREMENT,
+                checksum TEXT NOT NULL,
```
Comment (GiriDandu) [inline src/main/groovy/db/migration/V8__create_configs_table.groovy]: Why not make checksum VARCHAR(512) instead of TEXT?

File: src/main/groovy/db/migration/V8__create_configs_table.groovy [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+                deletedAt TIMESTAMP NULL,
+                PRIMARY KEY (id)
+            );
+        """
+
+        /*
+         * Add checksum column to flows table
+         */
+        commands += """
+            ALTER TABLE flows ADD COLUMN checksum TEXT;
```
Comment (GiriDandu) [inline src/main/groovy/db/migration/V8__create_configs_table.groovy]: If you do change checksum data type in modelConfigs table, make sure you change it flows table also

File: src/main/groovy/deploydb/ConfigReloadTask.groovy [hunk 11]
```diff
@@ -31,23 +28,15 @@ class ConfigReloadTask extends Task {
...
-        Session session = this.workFlow.deployDBApp.getSessionFactory().openSession()
-        try {
-            ManagedSessionContext.bind(session)
-            Transaction transaction = session.beginTransaction()
+        this.workFlow.deployDBApp.withHibernateSession() {
             try {
-                this.workFlow.loadConfigModels(true)
-                transaction.commit()
+                this.workFlow.loadConfigModels()
                 output.println("Done!")
```
Comment (GiriDandu) [inline src/main/groovy/deploydb/ConfigReloadTask.groovy]: can you use standard logging instead of stdout. Also "Done" is not very descriptive for logging

File: src/main/groovy/deploydb/ConfigReloadTask.groovy [hunk 12]
```diff
@@ -31,23 +28,15 @@ class ConfigReloadTask extends Task {
...
+            } catch (Exception e) {
+                logger.error("failed to reload the config with an exception: ", e)
+                output.println("Failed: " + e.getMessage())
```
Comment (GiriDandu) [inline src/main/groovy/deploydb/ConfigReloadTask.groovy]: Remove this output.println()

File: src/main/groovy/deploydb/WorkFlow.groovy [hunk 13]
```diff
@@ -59,21 +64,58 @@ public class WorkFlow {
...
+            filenames.each() { String filename ->
+                File modelFile = new File(modelDirectory, filename)
                 try {
-                    c.call(modelFile)
+                    /** Read YAML file into model object */
+                    def model = loader.load(modelFile)
+
+                    /** If registry is available, then insert model object into registry */
+                    String ident = defaultIdent
+                    if (registry) {
```
Comment (GiriDandu) [inline src/main/groovy/deploydb/WorkFlow.groovy]: When will the registry not present in this function?

File: src/cucumber/groovy/step_definitions/DeploymentSteps.groovy (hunk 5 above)
Comment (maheshkelkar) [inline src/cucumber/groovy/step_definitions/DeploymentSteps.groovy]: Yes

File: src/main/groovy/deploydb/WorkFlow.groovy [hunk 14]
```diff
@@ -189,31 +221,53 @@ public class WorkFlow {
...
 
-                tmpWebhook = this.webhookLoader.load(modelFile)
-                logger.debug("Loaded webhooks model from: ${modelFile.name}")
+                logger.debug("Loaded webhook model")
 
-                /* Now that we have found a valid webhook, we are done */
+                /* Now that we have found a valid global webhook, we are done */
                 throw new BreakLoopException()
             }
         } catch (BreakLoopException e) {
```
Comment (GiriDandu) [inline src/main/groovy/deploydb/WorkFlow.groovy]: Why there is no catch (all) here?

File: src/main/groovy/deploydb/WorkFlow.groovy [hunk 15]
```diff
@@ -189,31 +221,53 @@ public class WorkFlow {
...
+        }
+        if (this.deployDBApp.configChecksum == newConfigChecksum) {
+            logger.info("Ignoring as no change in DeployDB Model Configuration detected")
+            return
+        }
+        this.deployDBApp.configChecksum = newConfigChecksum
+
+        /* At least one service MUST be configured for deployDb to function properly */
+        if (tmpServiceRegistry.getAll().isEmpty()) {
+            logger.info("NO SERVICES ARE CONFIGURED. DeployDB would not function properly")
```
Comment (GiriDandu) [inline src/main/groovy/deploydb/WorkFlow.groovy]: I think we should move this log category to at least WARN or higher.  We want to get the operator attention here

File: src/cucumber/groovy/step_definitions/WebhookSteps.groovy (hunk 6 above)
Comment (maheshkelkar) [inline src/cucumber/groovy/step_definitions/WebhookSteps.groovy]: * webhook object reference is moved from WebhookManager to WorkFlow, hence we dont need it
* This particular cucumber "step" is used in used in a scenarios-3 in config.feature (create flow, reload config, trigger deployment STARTED). Hence, we are storing the webhook and environment in the ModelConfig

File: src/main/groovy/deploydb/WorkFlow.groovy [hunk 16]
```diff
@@ -221,7 +275,72 @@ public class WorkFlow {
...
+         * rebuild objects from DB
+         */
+        if (deployment.flow.configChecksum != this.deployDBApp.configChecksum) {
+            /* Load webhook from config */
+            models.ModelConfig webhookConfig =
+                    this.modelConfigDAO.findModelConfig(ModelType.WEBHOOK,
+                            defaultIdent, deployment.flow.configChecksum)
+            if (webhookConfig) {
+                models.Webhook.Webhook fetchedWebhook =
+                        this.webhookLoader.loadFromString(webhookConfig.contents)
```
Comment (GiriDandu) [inline src/main/groovy/deploydb/WorkFlow.groovy]: Are we certain that this won't throw an exception?

File: src/cucumber/groovy/step_definitions/WebhookSteps.groovy (hunk 7 above)
Comment (maheshkelkar) [inline src/cucumber/groovy/step_definitions/WebhookSteps.groovy]: * This is Intellij's doing
* If I had done it, I am not going to put these in-consequential but appropriate changes stated in the PR ;-)

File: src/cucumber/groovy/step_definitions/WebhookSteps.groovy (hunk 8 above)
Comment (maheshkelkar) [inline src/cucumber/groovy/step_definitions/WebhookSteps.groovy]: ok

File: src/main/groovy/db/migration/V8__create_configs_table.groovy (hunk 9 above)
Comment (maheshkelkar) [inline src/main/groovy/db/migration/V8__create_configs_table.groovy]: I agree this needs to be changed. There is inconsistency in use of VARCHAR(8192) or TEXT everywhere in the code. If I had used VARCHAR, why not TEXT, was a valid question to. 

Let me investigate if there is a limit on output of md5hex() output, if yes, I will put it to that size OR just change it to VARCHAR

File: src/main/groovy/db/migration/V8__create_configs_table.groovy (hunk 10 above)
Comment (maheshkelkar) [inline src/main/groovy/db/migration/V8__create_configs_table.groovy]: ok

File: src/main/groovy/deploydb/ConfigReloadTask.groovy (hunk 11 above)
Comment (maheshkelkar) [inline src/main/groovy/deploydb/ConfigReloadTask.groovy]: This "output" is return field for admin "Task". This was reviewed as a part of #120. If it needs to be addressed, we can evaluate it later

File: src/main/groovy/deploydb/ConfigReloadTask.groovy (hunk 12 above)
Comment (maheshkelkar) [inline src/main/groovy/deploydb/ConfigReloadTask.groovy]: We can't, this is a response back to the "ConfigReload" task

File: src/main/groovy/deploydb/WorkFlow.groovy (hunk 13 above)
Comment (maheshkelkar) [inline src/main/groovy/deploydb/WorkFlow.groovy]: In case of webhook, the registry is not passed, because its singular. In that case "defaultIdent" is used.

File: src/main/groovy/deploydb/WorkFlow.groovy (hunk 14 above)
Comment (maheshkelkar) [inline src/main/groovy/deploydb/WorkFlow.groovy]: This was a part of #44. @rtyler wanted it to be handled by the caller.

File: src/main/groovy/deploydb/WorkFlow.groovy (hunk 15 above)
Comment (maheshkelkar) [inline src/main/groovy/deploydb/WorkFlow.groovy]: This code is unchanged (simply moved around due to new checksum calculations). But, WARN is too strong to do so.

File: src/main/groovy/deploydb/WorkFlow.groovy (hunk 16 above)
Comment (maheshkelkar) [inline src/main/groovy/deploydb/WorkFlow.groovy]: I don't expect any known exceptions. Similar to other processing, it will abort the process and handled by the caller (i.e. resource in most of the cases)

**Comment by maheshkelkar** [comment]
Hi @mbbx6spp, could you please review the schema changes in this commit

**Comment by rtyler** [comment]
Looks fine to me but needs to be synced with master looks like

**Comment by GiriDandu** [comment]
:+1:
